scanner/services/converters/converters.go

```diff
@@ -0,0 +1,400 @@
+	if len(ccPkgVulnerabilities) > 0 {
+		pkgVulns = make(map[string]*v4.StringList, len(ccPkgVulnerabilities))
+	}
+	for k, v := range ccPkgVulnerabilities {
+		if v == nil {
+			continue
+		}
+		pkgVulns[k] = &v4.StringList{
+			Values: append([]string(nil), v...),
+		}
+	}
+	return pkgVulns
+}
+
+func toProtoV4VulnerabilitiesMap(vulns map[string]*claircore.Vulnerability) (map[string]*v4.VulnerabilityReport_Vulnerability, error) {
+	if vulns == nil {
+		return nil, nil
+	}
+	var vulnerabilities map[string]*v4.VulnerabilityReport_Vulnerability
+	if len(vulns) > 0 {
+		vulnerabilities = make(map[string]*v4.VulnerabilityReport_Vulnerability, len(vulns))
+	}
+	for k, v := range vulns {
+		if v == nil {
+			continue
+		}
+		issued, err := types.TimestampProto(v.Issued)
+		if err != nil {
+			return nil, err
+		}
+		var pkgID string
+		if v.Package != nil {
+			pkgID = v.Package.ID
+		}
+		var distID string
+		if v.Dist != nil {
+			distID = v.Dist.ID
+		}
+		var repoID string
+		if v.Repo != nil {
+			repoID = v.Repo.ID
+		}
+		normalizedSeverity := toProtoV4VulnerabilitySeverity(v.NormalizedSeverity)
+		vulnerabilities[k] = &v4.VulnerabilityReport_Vulnerability{
+			Id:                 v.ID,
+			Name:               v.Name,
+			Description:        v.Description,
+			Issued:             issued,
+			Link:               v.Links,
+			Severity:           v.Severity,
+			NormalizedSeverity: normalizedSeverity,
+			PackageId:          pkgID,
+			DistributionId:     distID,
+			RepositoryId:       repoID,
+			FixedInVersion:     v.FixedInVersion,
+		}
+	}
+	return vulnerabilities, nil
+}
+
+func toProtoV4VulnerabilitySeverity(ccSeverity claircore.Severity) v4.VulnerabilityReport_Vulnerability_Severity {
+	if mappedSeverity, ok := severityMapping[ccSeverity]; ok {
+		return mappedSeverity
+	}
+	return v4.VulnerabilityReport_Vulnerability_SEVERITY_UNSPECIFIED
+}
+
+func toCPEString(c cpe.WFN) string {
+	return c.BindFS()
+}
+
+func toDigestString(digest claircore.Digest) string {
+	return digest.String()
+}
+
+// convertMapToSlice converts generic maps keyed by strings to a slice using a
+// provided conversion function.
+func convertMapToSlice[IN any, OUT any](convF func(*IN) *OUT, in map[string]*IN) (out []*OUT) {
+	for _, i := range in {
+		out = append(out, convF(i))
+	}
+	return out
+}
+
+func toClairCoreCPE(s string) (cpe.WFN, error) {
+	c, err := cpe.UnbindFS(s)
+	if err != nil {
+		return c, fmt.Errorf("%q: %s", s, strings.TrimPrefix(err.Error(), "cpe: "))
+	}
+	return c, nil
+}
+
+func toClairCorePackage(p *v4.Package) (string, *claircore.Package, error) {
+	// Conversion functions.
+	toNormalizedVersion := func(v *v4.NormalizedVersion) (ccV claircore.Version) {
+		ccV.Kind = v.GetKind()
+		copy(ccV.V[:], v.GetV())
+		return
+	}
+	toSourcePackage := func(src *v4.Package) (*claircore.Package, error) {
+		if src == nil {
+			return nil, nil
+		}
+		// Prevent deeper recursion.
+		if src.GetSource() != nil {
+			src.Source = nil
```

Comment:
Same comment as above regarding converter modifying the input object - perhaps should avoid if possible.